### PR TITLE
feat(frontend): support single dimension array for extended-mode query

### PR DESCRIPTION
**Pull request by jetjinser**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Split `e2e_test/batch/types/array_ty.slt.part` into 1-dim `e2e_test/batch/types/array_1d_ty.slt.part` and multi-dim (keep original name).

Partially resolved - #7949.

```python
import psycopg

def ep(query):
    cur.execute(query, binary=True)  # type: ignore
    for row in cur:
        print(row)
    print()

conn = psycopg.connect(host="localhost", port="4566", dbname="dev", user="root")
with conn.cursor() as cur:
    int_tys = ["int2", "int4", "int8", "rw_int256", "numeric"]
    flt_tys = ["float4", "float8"]
    txt_tys = ["text", "bytea"]

    for it in int_tys:
        query = f"select Array[1::{it}, 2::{it}, 3::{it}]"
        ep(query)

    for ft in flt_tys:
        query = f"select Array[1.1::{ft}, 2.2::{ft}, 3.3::{ft}]"
        ep(query)

    for tt in txt_tys:
        query = f"select Array['a'::{tt}, 'b'::{tt}, 'c'::{tt}]"
        ep(query)

    # error: n-dim where n > 1
    # query = "select Array[Array[1, 2, 3], Array[4, 5, 6]]"
    # ep(query)
```

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

### CODE REVIEW

File: src/common/src/types/to_sql.rs [hunk 1]
```diff
@@ -74,3 +76,21 @@ impl ToSql for JsonbRef<'_> {
...
+        Self: Sized,
+    {
+        matches!(ty.kind(), postgres_types::Kind::Array(_))
+    }
+
+    fn to_sql(&self, ty: &Type, out: &mut BytesMut) -> Result<IsNull, Box<dyn Error + Sync + Send>>
+    where
+        Self: Sized,
+    {
+        self.iter().collect_vec().to_sql(ty, out)
```
Comment (xiangjinwu) [inline src/common/src/types/to_sql.rs:94]: Make sure only 1 dimension?

File: src/common/src/types/to_binary.rs [hunk 2]
```diff
@@ -87,14 +87,54 @@ impl ToBinary for ScalarRefImpl<'_> {
...
+            DataType::Decimal => Type::NUMERIC_ARRAY,
+            DataType::Date => Type::DATE_ARRAY,
+            DataType::Varchar => Type::VARCHAR_ARRAY,
+            DataType::Time => Type::TIME_ARRAY,
+            DataType::Timestamp => Type::TIMESTAMP_ARRAY,
+            DataType::Timestamptz => Type::TIMESTAMPTZ_ARRAY,
+            DataType::Interval => Type::INTERVAL_ARRAY,
+            DataType::Bytea => Type::BYTEA_ARRAY,
+            DataType::Jsonb => Type::JSONB_ARRAY,
+            DataType::Serial => Type::INT4_ARRAY,
```
[the remark this answers is not in the document]
Comment (jetjinser) [reply src/common/src/types/to_binary.rs:120]: I wrote it based on this [table](https://www.postgresql.org/docs/16/datatype.html#DATATYPE-TABLE). What it means is that serial(serial4) is a 4-byte signed Integer, but does not include negative numbers: https://www.postgresql.org/docs/16/datatype-numeric.html#DATATYPE-SERIAL.

File: src/common/src/types/to_sql.rs (hunk 1 above)
Comment (jetjinser) [reply src/common/src/types/to_sql.rs:94]: https://docs.rs/postgres-types/0.2.6/postgres_types/trait.ToSql.html#arrays

https://github.com/risingwavelabs/risingwave/blob/855cd7d9037fa0574d8c9dec4c8e2c1621c9f4f8/src/common/src/types/to_binary.rs#L124-L127
I think it should be guaranteed here that there will only be an array of 1 dimension?

File: src/common/src/types/to_binary.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiangjinwu) [reply src/common/src/types/to_binary.rs:120]: The (internal) `serial` type in RisingWave (#8392) is not the same as `serial` sugar in PostgreSQL.

File: src/common/src/types/to_sql.rs (hunk 1 above)
Comment (xiangjinwu) [reply src/common/src/types/to_sql.rs:94]: But as a general trait implementation, `to_sql` may have other callers (in the future) in addition to `to_binary`.

File: src/common/src/types/to_sql.rs [hunk 3]
```diff
@@ -74,3 +76,21 @@ impl ToSql for JsonbRef<'_> {
...
 }
+
+impl ToSql for ListRef<'_> {
+    to_sql_checked!();
+
+    fn accepts(ty: &Type) -> bool
+    where
+        Self: Sized,
+    {
+        matches!(ty.kind(), postgres_types::Kind::Array(_))
```
[the remark this answers is not in the document]
Comment (jetjinser) [reply src/common/src/types/to_sql.rs:87]: I don't think `accept` has a way to check the inner type... it only know the `postgres_type::Type`.
It should be a limitation of the `ToSql` api.

File: src/common/src/types/to_sql.rs (hunk 1 above)
Comment (xiangjinwu) [reply src/common/src/types/to_sql.rs:94]: The link above is for `FromSql`. There seems to be no check of multiple dimensions in `ToSql` there:
https://github.com/sfackler/rust-postgres/blob/e528e01bbe2f7e198a55112fd6de3a24db3b8a6e/postgres-types/src/lib.rs#L960

Can we confirm by actually executing `to_sql` on a `Vec<Vec<i32>>` or `ListRef` of `ListRef` of `i32` value?

File: src/common/src/types/to_sql.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jetjinser) [reply src/common/src/types/to_sql.rs]: Oh I understand, but we can still only check whether ty is a one-dimensional array. Is this implementation correct?

**Pull request closed without merging**